src/Http/WebUtilities/src/MultipartSectionPipeReader.cs

```diff
@@ -0,0 +1,600 @@
+{
+    public class MultipartSectionPipeReader
+    {
+        private readonly PipeReader _pipeReader;
+        private bool _finished = false;
+        private bool _metadataSkipped = false;
+        private readonly MultipartBoundary _boundary;
+        private int _partialMatchIndex = 0;
+        private const int StackAllocThreshold = 128;
+        private ReadOnlySequence<byte> _buffer;
+
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+        private static ReadOnlySpan<byte> EndOfFileDelimiter => new byte[] { (byte)'-', (byte)'-' };
+
+
+        public bool CanSeek { get; }
+
+        public long RawLength { get; private set; } = 0;
+
+        public long? LengthLimit { get; private set; }
+        public long Length { get; private set; } = 0;
+
+        internal MultipartSectionPipeReader(PipeReader pipeReader, MultipartBoundary boundary, bool canSeek)
+        {
+            _pipeReader = pipeReader;
+            _boundary = boundary;
+            CanSeek = canSeek;
+        }
+
+        public int Read(byte[] buffer, int offset, int count)
+        {
+            if (_finished && _metadataSkipped)
+            {
+                return 0;
+            }
+            int consumed = 0;
+            ReadOnlySequence<byte> sequence = default;
+            ReadResult readResult = default;
+            while (true)
+            {
+                if (!sequence.IsEmpty)
+                {
+                    if (!_finished)
+                    {
+                        var (didReachEnd, copied) = TryCopyToEnd(ref sequence, buffer, offset, count, readResult.IsCompleted);
+                        _pipeReader.AdvanceTo(sequence.Start);
+                        consumed += copied;
+                        if (didReachEnd)
+                        {
+                            _finished = true;
+                            Length += copied;
+                            RawLength += copied;
+                        }
+
+                        if (copied > buffer.Length - offset)
+                        {
+                            Length += copied;
+                            RawLength += copied;
+                            return consumed;
+                        }
+                        offset += copied;
+                    }
+                    else if (!_metadataSkipped)
+                    {
+                        if (TrySkipMetadata(ref sequence, readResult.IsCompleted))
+                        {
+                            _pipeReader.AdvanceTo(sequence.Start);
+                            _metadataSkipped = true;
+                            return consumed;
+                        }
+                    }
+                    else
+                    {
+                        return consumed;
+                    }
+                }
+
+                if (_pipeReader.TryRead(out readResult))
+                {
+                    if (readResult.IsCompleted)
```

Comment:
This doesn't look right. The readResult can complete and return data at the same time. That buffer needs to be parsed before you can claim your finished. Also, it should error if it finishes without finding the final boundary.